lstchain/io/event_selection.py

```diff
@@ -64,13 +77,105 @@ class DL3FixedCuts(Component):
         default_value=[1],
     ).tag(config=True)
 
-    def gh_cut(self, data):
-        return data[data["gh_score"] > self.fixed_gh_cut]
+    def apply_global_gh_cut(self, data):
+        """
+        Applying a global gammaness cut on a given data
+        """
+        return data[data["gh_score"] > self.global_gh_cut]
+
+    def energy_dependent_gh_cuts(
+        self, data, energy_bins, min_value=0.1,
+        max_value=0.99, smoothing=None, min_events=10
+    ):
+        """
+        Evaluating energy-dependent gammaness cuts, in a given
+        data, with provided reco energy bins, and other parameters to
+        pass to the pyirf.cuts.calculate_percentile_cut function
+        """
+        # To be sure to have the efficiency in between 0 and 100
+        if self.gh_efficiency < 1:
+            self.gh_efficiency *= 100
+
+        gh_cuts = calculate_percentile_cut(
+            data["gh_score"],
+            data["reco_energy"],
+            bins=energy_bins,
+            min_value=min_value,
+            max_value=max_value,
+            fill_value=data["gh_score"].max(),
+            percentile=100 - self.gh_efficiency,
+            smoothing=smoothing,
+            min_events=min_events,
+        )
+        return gh_cuts
 
-    def theta_cut(self, data):
-        return data[data["theta"].to_value(u.deg) < self.fixed_theta_cut]
+    def apply_energy_dependent_gh_cuts(self, data, gh_cuts):
+        """
+        Applying a given energy-dependent gh cuts to a data file, along the reco
+        energy bins provided.
+        """
+
+        data["selected_gh"] = evaluate_binned_cut(
+            data["gh_score"],
+            data["reco_energy"],
+            gh_cuts,
+            operator.ge,
+        )
+        return data[data["selected_gh"]]
+
+    def apply_global_theta_cut(self, data):
+        """
+        Applying a global theta cut on a given data
+        """
+        return data[data["theta"].to_value(u.deg) < self.global_theta_cut]
+
+    def energy_dependent_theta_cuts(
+        self, data, energy_bins, min_value=0.05 * u.deg,
+        fill_value=0.32 * u.deg, max_value=0.32 * u.deg,
+        smoothing=None, min_events=10
+    ):
+        """
+        Evaluating an optimized energy-dependent theta cuts, in a given
+        data, with provided reco energy bins, and other parameters to
+        pass to the pyirf.cuts.calculate_percentile_cut function.
+
+        Note: Using too fine binning will result in too un-smooth cuts.
+        """
+        # To be sure to have the efficiency in between 0 and 100
+        if self.theta_containment < 1:
+            self.theta_containment *= 100
```

Comment:
Same comment as before. Does not seem a reasonable thing to do, and if you want to spot a suspicious input setting (too low efficiency or containment) I would do it upon reading the configuration file, not in the function which calculates the cuts.